plutus-core/plutus-ir/src/PlutusIR/Transform/RewriteRules/DecodeEncodeUtf8.hs

```diff
@@ -0,0 +1,32 @@
+{-# LANGUAGE GADTs           #-}
+{-# LANGUAGE LambdaCase      #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE TypeOperators   #-}
+
+module PlutusIR.Transform.RewriteRules.DecodeEncodeUtf8
+  ( decodeEncodeUtf8
+  ) where
+
+
+import PlutusCore.Default
+import PlutusIR
+
+{- | This rewrites: `(decodeUtf8 (encodeUtf8 x)) => x`
+
+This rewrite is safe because x is either valid by construction *In Haskell, but not in GHC-CORE*
+(see <https://hackage.haskell.org/package/text-2.1/docs/Data-Text.html#g:2>)
+, or x is bottom.
+
+The opposite rewrite `(encodeUtf8 (decodeUtf8 x)) => x` is definitely not safe.
+-}
+-- FIXME: Unfortunately, unicode text is currently broken (at least on plutus-tx level), so
+-- we disable this rewrite until fix is in and further tested. See PLT-8314
+decodeEncodeUtf8 :: (Semigroup a, t ~ Term tyname name uni DefaultFun a) => t -> t
+decodeEncodeUtf8 = \case
+    BA DecodeUtf8 a1 a2 (BA EncodeUtf8 a3 a4 t) ->
+        -- place the missed annotations inside the rewritten term
+        (<> a1 <> a2 <> a3 <> a4) <$> t
```

Comment:
Just nitpicking, but I'd use `sconcat :: NonEmpty a -> a` instead, because we have

```
data Provenance a = 
                  <...>
                  | MultipleSources (S.Set (Provenance a))
```

and that would be a natural target for `sconcat`.

... except no implementation for `sconcat` is provided in

```haskell
instance Ord a => Semigroup (Provenance a) where
    x <> y = MultipleSources (toSet x `S.union` toSet y)
      where
        toSet = \case
            MultipleSources ps -> ps
            other              -> S.singleton other
```

but I think it's just an oversight and so I'd provide an implementation for `sconcat` and define `<>` in terms of that instead.

Do feel free to ignore this comment.